Rework the docs for NEWS/What's New entries (#1238)
python/devguide: core-developers/committing.rst

Updating NEWS and What's New in Python
--------------------------------------

> Changes that entail NEWS entries

Reviewer: ```suggestion
Changes that require NEWS entries
```
I would use simpler terms here.  I think "require" still fits even if the section also describes what changes *don't* require a NEWS entry.

Author: Sure, I'm fine with using 'require'.

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Almost all changes made to the code base deserve an entry in ``Misc/NEWS.d``.

Reviewer: FWIW this is roughly the same wording used by `bedevere`, and during the last sprint we had a discussion about it and ended up tweaking it:
* https://github.com/python/bedevere/issues/439
* https://github.com/python/bedevere/pull/447
* https://github.com/python/bedevere/pull/593

Author: Thanks for those links, Ezio. We could align it with the wording used by Bedevere, but I think the current wording is fine.

> There are two notable exceptions to this general principle, and they

Reviewer: I would say that minor changes (like typo fixes, some changes to documentation and tests, etc) are another notable exception.  These are mentioned below, but I find it a bit weird that it mentions two "notable" exception (that are somewhat uncommon) and then adds more exception afterwards.

IOW, I would just say something like: "NEWS entries are not needed for: <list of exceptions>"

Author: Yeah, I think the text is easier if we collect the exceptions in a single place. I'll give it a go.

Author: How about this (diff against PR):

```diff
diff --git a/core-developers/committing.rst b/core-developers/committing.rst
index 8d5e2d4..4e11ed5 100644
--- a/core-developers/committing.rst
+++ b/core-developers/committing.rst
@@ -83,11 +83,17 @@ Updating NEWS and What's New in Python
 Changes that require NEWS entries
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-Almost all changes made to the code base deserve an entry in ``Misc/NEWS.d``.
-There are two notable exceptions to this general principle, and they
-both relate to changes that already have a ``NEWS`` entry,
-or changes that have not yet been included in any formal release
-(including alpha and beta releases):
+Most changes made to the code base deserve an entry in :cpy-file:`Misc/NEWS.d`,
+except for the following:
+
+* documentation changes in general
+* test changes in general
+* strictly internal changes with no user-visible effects
+* changes that already have a ``NEWS`` entry
+* changes that have not yet been included in any formal release
+  (including alpha and beta releases)
+
+For the latter two exceptions, note the following:
 
 #. **If a change is reverted prior to release**, then the corresponding
    entry is simply removed. Otherwise, a new entry must be added noting
```

Reviewer 2: That sounds great!

Reviewer 3: Much better. Let's use "last two" instead of "latter two".

https://www.merriam-webster.com/grammar/former-and-latter-usage

Remove "in general" from:

```
* documentation changes in general
* test changes in general
```

---

This one is confusing, does it more specifically mean _reverts_ rather than _changes_?

```
* changes that have not yet been included in any formal release
  (including alpha and beta releases)
```

Updating NEWS and What's New in Python
--------------------------------------

Changes that require NEWS entries
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Most changes made to the code base deserve an entry in :cpy-file:`Misc/NEWS.d`,

Reviewer 3: ```suggestion
Most changes made to the codebase deserve an entry in :cpy-file:`Misc/NEWS.d`,
```

except for the following:

> * documentation changes in general
> * test changes in general

Author: cc. @hugovk:

```suggestion
* documentation changes
* test changes
```

... or possibly:

```suggestion
* most documentation changes
* most test changes
```

* strictly internal changes with no user-visible effects
* changes that already have a ``NEWS`` entry

> * changes that have not yet been included in any formal release

Author: Continuing the discussion from https://github.com/python/devguide/pull/1238#discussion_r1417664126.

I pretty much just reused the original wording, but I agree it is slightly confusing. It helps with the explanation a few lines below, though.

Author: I think "reverts" is ok, so I'm inclined to apply Hugo's suggestion. For anyone familiar with Git, it makes sense as a noun (`git commit`, `git revert`). I think Ezio's suggestion reads a little off-kilter (perhaps because of the double "being").

The idea of a bad/good list could work out, but I'd prefer if we could do that in a follow-up PR.

Thanks, both of you.

Author: Let's go for "revert" for now. There's always room for further improvements in future PRs :)

  (including alpha and beta releases)

> For the latter two exceptions, note the following:

Author: ```suggestion
For the last two, note the following:
```

   change and the original** ``NEWS`` **entry remains valid**, then no additional
   entry is needed.

Other changes that generally do not require ``NEWS`` entries are documentation changes,
test changes, and strictly internal changes with no user-visible effects.

> Changes that entail What's New in Python entries

Reviewer: ```suggestion
Changes that require What's New in Python entries
```

```suggestion
Changes that require "What's New in Python" entries
```

Same as above, possibly with quotes.

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

If a change is particularly interesting for end users (e.g. new features,
significant improvements, or backwards-incompatible changes), an entry in

> the ``What's New in Python`` document (in ``Doc/whatsnew/``) should be added

Reviewer: ```suggestion
the ``What's New in Python`` document (in :cpy-file:`Doc/whatsnew/`) should be added
```
(this works for dirs too IIRC)

   change and the original** ``NEWS`` **entry remains valid**, then no additional
   entry is needed.

Other changes that generally do not require ``NEWS`` entries are documentation changes,
test changes, and strictly internal changes with no user-visible effects.

> 

Author: Thanks!

As a result, a file name can look something like
``Misc/NEWS.d/next/Library/2017-05-27-16-46-23.gh-issue-12345.Yl4gI2.rst``.

How to write a NEWS entry
^^^^^^^^^^^^^^^^^^^^^^^^^

All ``NEWS`` entries end up being part of the changelog.
The changelog contains *a lot* of entries,
and its intended audience is mainly users, not core devs and contributors.
Take this into consideration when wording your ``NEWS`` entry.
Describe the user-visible effects of your change succinctly and accurately;
avoid long technical elaborations, digressions, and do not expect or require

> the reader to have read the actuall diff for the change.

Reviewer: ```suggestion
the reader to have read the actual diff for the change.
```

Author: Oops, thanks!

find more information. You can build HTML and verify that the
link target is appropriate by using :ref:`make html <building-using-make>`.

While Sphinx roles can be beneficial to readers, they are not required.

> Inline ````code blocks```` can be used instead.

Author: I deliberately removed this. The above sentence says "*should* be valid reST" (emphasis mine), so IMO we do not need to repeat that it is not required. Moreover, in most cases, we actually ask contributors explicitly add markup.
